test/content-length.js

```diff
@@ -329,3 +288,129 @@ test('request streaming with Readable.from(buf)', (t) => {
     })
   })
 })
+
+test('request DELETE, content-length=0, with body', (t) => {
+  t.plan(3)
+  const server = createServer((req, res) => {
+    res.end()
+  })
+  t.teardown(server.close.bind(server))
+  server.listen(0, () => {
+    const client = new Client(`http://localhost:${server.address().port}`)
+    t.teardown(client.destroy.bind(client))
+
+    client.request({
+      path: '/',
+      method: 'DELETE',
+      headers: {
+        'content-length': 0
+      },
+      body: new Readable({
+        read () {
+          this.push('asd')
+          this.push(null)
+        }
+      }),
+    }, (err) => {
+      t.type(err, errors.RequestContentLengthMismatchError)
+    })
+
+    client.request({
+      path: '/',
+      method: 'DELETE',
+      headers: {
+        'content-length': 0
+      }
+    }, (err) => {
+      t.error(err)
+    })
+
+    client.on('disconnect', () => {
+      t.pass()
+    })
+  })
+})
+
+test('content-length shouldSendContentLength=false', (t) => {
+  t.plan(9)
+  const server = createServer((req, res) => {
+    res.end()
```

Comment:
Can we verify that the `content-length` is `undefined` or `0` in any case?

Reply:
What about the `request` being handled at the server; shouldn't we test that also its `0`?

Reply:
i see what you're saying. test added for both this and the DELETE below.

Do we want to strictly NOT send content-length header for GET/HEAD/OPTION requests even if the request contains body? This most likely means we'll have to check `shouldSendContentLength` here as well https://github.com/nodejs/undici/blob/741b8118e1452d9c00358b2c33d7068683e9dcc6/lib/client.js#L1641

Reply:
Hmm, no you are right that will be a breaking change and we should not need that. Let's go with your suggestion then, of adding an extra guard for the `content-length`, in that way we keep the current semantics and APIs valid 👍

Let's try adding your changes, and see if related; `redirect-pipeline` is flaky, but worth it check if with the changes fixes

Reply:
Agreed, i've backed out the last batch of changes.

I think I'm going to leave things as is for backwards compat sake.

From testing it seems like forcing `content-length: 0` on `shouldSendContentLength=false` methods will also force any body written via `socket.write(body)` to be empty, this will cause tests in `test/client-post.js` to fail.

I think this is a decision for another PR? @metcoder95

Reply:
Yeah, can you open an issue based on that; it will be worth it to chat as it will be a breaking change for sure 👍

Reply:
For sure, https://github.com/nodejs/undici/issues/2357